### PR TITLE
Update pipeline

### PR DESCRIPTION
**Pull request by jeremybumsted**
Bring the pipeline configuration up to date by
- moving pipeline into a YAML config from a shell script (pipeline.sh can be removed after verified)
- updating docker-compose to use the correct images
- updating the version to 0.2.2 for linter checks

Added some additional bits (including requested changes)
- Fixed integration tests
- Updated out-of-date docker images used in integration tests so they pass now
- Fixed BATS tests relevant to current functionality, skipping many that will pass in future PRs
- Updated plugin schema

### CODE REVIEW

**Comment by jeremybumsted** [comment]
This PR will update the pipeline so we have the framework to run the pipeline in BK, but it's expected to fail.

File: .buildkite/pipeline.yml [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+  SVC_VAULT_VERSION: "1.11.2" # this probably needs to get bumped over time
+  VAULT_ADDR: "http://vault-svc:8200"
+  BUILDKITE_PIPELINE_SLUG: "my_pipeline"
+  DUMP_ENV: "true"
+  VAULT_DEV_ROOT_TOKEN_ID: "88F4384B-98E9-4AE3-B00C-F55678F89080"
+steps:
+  - label: ":sparkles: Lint"
+    plugins:
+      - docker-compose#v4.5.0:
+          run: lint
```
Comment (toote) [inline .buildkite/pipeline.yml:13]: why not use the linter plugin instead?

File: .buildkite/pipeline.yml [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+
+  - label: ":shell: Shellcheck"
+    plugins:
+      - docker-compose#v4.5.0:
+          run: shellcheck
```
Comment (toote) [inline .buildkite/pipeline.yml:18]: why not use the shellcheck plugin instead?

File: plugin.yml [hunk 3]
```diff
@@ -22,7 +22,4 @@ configuration:
   required:
     - server
     - auth
-  additionalProperties: false
-  dependencies:
-    auth: [ method ]
```
Comment (toote) [inline plugin.yml:27]: why these changes?

File: tests/auth-tests.bats [hunk 4]
```diff
@@ -2,10 +2,10 @@
...
 load '/usr/local/lib/bats/load.bash'
 
-# export SSH_AGENT_STUB_DEBUG=/dev/tty
-# export SSH_ADD_STUB_DEBUG=/dev/tty
-# export VAULT_STUB_DEBUG=/dev/tty
-# export GIT_STUB_DEBUG=/dev/tty
+export SSH_AGENT_STUB_DEBUG=/dev/tty
+export SSH_ADD_STUB_DEBUG=/dev/tty
+export VAULT_STUB_DEBUG=/dev/tty
+export GIT_STUB_DEBUG=/dev/tty
```
Comment (toote) [inline tests/auth-tests.bats:8]: these changes should not be committed to the repository

File: .buildkite/pipeline.yml [hunk 5]
```diff
@@ -0,0 +1,40 @@
+env:
+  BK_QUEUE_NAME: "testqueue" # probably don't need this
+  TESTER_VAULT_VERSION: "1.11.2" # this probably needs to get bumped over time
+  SVC_VAULT_VERSION: "1.11.2" # this probably needs to get bumped over time
+  VAULT_ADDR: "http://vault-svc:8200"
+  BUILDKITE_PIPELINE_SLUG: "my_pipeline"
+  DUMP_ENV: "true"
+  VAULT_DEV_ROOT_TOKEN_ID: "88F4384B-98E9-4AE3-B00C-F55678F89080"
```
Comment (toote) [inline .buildkite/pipeline.yml:8]: if they are only used in a single step, why not add them on that step instead?

File: .buildkite/pipeline.yml [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+      - docker-compose#v4.5.0:
+          run: shellcheck
+
+  - label: ":bat: :test_tube: BATS tests"
+    plugins:
+      - docker-compose#v4.5.0:
+          run: tests
+          
+  - label: ":vault: :test_tube: Integration Tests"
+    command:  /app/.buildkite/steps/test_envvar.sh
```
Comment (toote) [inline .buildkite/pipeline.yml:26]: it may be a better idea to make this command relative and set the `workdir` in the plugin

File: .buildkite/pipeline.yml [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+        run: vault-tester
+        env:
+          - VAULT_ADDR=\${VAULT_ADDR}
+          - BUILDKITE_PLUGIN_VAULT_SECRETS_SERVER=\${VAULT_ADDR}
+          - BUILDKITE_PIPELINE_SLUG=\${BUILDKITE_PIPELINE_SLUG}
+          - BUILDKITE_PLUGIN_VAULT_SECRETS_DUMP_ENV=\${DUMP_ENV}
+          - VAULT_DEV_ROOT_TOKEN_ID=\${VAULT_DEV_ROOT_TOKEN_ID}
+          - BUILDKITE_PLUGIN_VAULT_SECRETS_AUTH_TOKEN=\${VAULT_DEV_ROOT_TOKEN_ID}
+          - SVC_VAULT_VERSION=\${SVC_VAULT_VERSION}
+          - TESTER_VAULT_VERSION=\${TESTER_VAULT_VERSION}
```
Comment (toote) [inline .buildkite/pipeline.yml:40]: Most of these variables are already defined in the docker compose file so I don't even think they are really necessary to specify here

**Review by toote — CHANGES_REQUESTED** [review]
A few minor suggestions and a big one (with regards to the uncommented debug lines in the tests).

That said, I believe that there is a bigger issue at hand: the testing of the plugin runs a very weird setup with an image we don't know where it comes from. My hunch is that it should be possible to run the test in the same vault image as the only requisite appears to be that vault client is installed... in which case I'm not sure that we need a separate vault service docker container

**Comment by jeremybumsted** [comment]
Thanks for the review @toote! I agree with all of your suggestions 😅 I'll get those changes in sometime this week!

> That said, I believe that there is a bigger issue at hand: the testing of the plugin runs a very weird setup with an image we don't know where it comes from. My hunch is that it should be possible to run the test in the same vault image as the only requisite appears to be that vault client is installed... in which case I'm not sure that we need a separate vault service docker container

I believe this image runs a vault server in dev mode, so that the integration tests can be run without the need for an actual vault server. I wanted to address this in a different PR, but it might make more sense to have it dealt with in here too.

File: .buildkite/pipeline.yml (hunk 7 above)
Comment (jeremybumsted) [reply .buildkite/pipeline.yml]: Agreed - cleaned this up.

File: .buildkite/pipeline.yml (hunk 5 above)
Comment (jeremybumsted) [reply .buildkite/pipeline.yml]: I tried defining the values a few different ways (most of these values are also defined or have defaults in the docker-compose) with varying degrees of success. Ended up only leaving `VAULT_DEV_ROOT_TOKEN_ID` and `VAULT_ADDR` as globally defined, the rest are defined in the plugin env directly.

File: plugin.yml (hunk 3 above)
Comment (jeremybumsted) [reply plugin.yml]: Fixed this in a more recent commit, we want `auth` to be a required property.

File: docker-compose-integration.yml [hunk 8]
```diff
@@ -1,21 +1,22 @@
...
-       test: ["CMD", "vault", "status"]
-       interval: 30s
-       timeout: 10s
-       retries: 5
     environment:
       - VAULT_DEV_ROOT_TOKEN_ID=${VAULT_DEV_ROOT_TOKEN_ID}
       - VAULT_TOKEN=${VAULT_DEV_ROOT_TOKEN_ID}
-      - VAULT_ADDR=${VAULT_ADDR}
-      - VPC_PROXY=${VPC_PROXY}
+      - VPC_PROXY=${VPC_PROXY:-}
```
Comment (toote) [inline docker-compose-integration.yml:15]: this is the same as:

```suggestion
      - VPC_PROXY
```

File: docker-compose-integration.yml [hunk 9]
```diff
@@ -1,21 +1,22 @@
...
+      retries: 5
     cap_add:
     - IPC_LOCK
-    healthcheck:
-       test: ["CMD", "vault", "status"]
-       interval: 30s
-       timeout: 10s
-       retries: 5
     environment:
       - VAULT_DEV_ROOT_TOKEN_ID=${VAULT_DEV_ROOT_TOKEN_ID}
```
Comment (toote) [inline docker-compose-integration.yml:13]: this is the same as:

```suggestion
      - VAULT_DEV_ROOT_TOKEN_ID
```

File: docker-compose-integration.yml [hunk 10]
```diff
@@ -26,13 +27,13 @@ services:
     cap_add:
       - IPC_LOCK
     environment:
-      - VAULT_DEV_ROOT_TOKEN_ID=${VAULT_DEV_ROOT_TOKEN_ID}
-      - VAULT_ADDR=${VAULT_ADDR}
-      - VPC_PROXY=${VPC_PROXY}
+      - VAULT_DEV_ROOT_TOKEN_ID
+      - VAULT_ADDR=${VAULT_ADDR:-}
+      - VPC_PROXY=${VPC_PROXY:-}
```
Comment (toote) [inline docker-compose-integration.yml:32]: this is the same as:
```suggestion
      - VAULT_ADDR
      - VPC_PROXY
```

File: lib/shared.bash [hunk 11]
```diff
@@ -1,11 +1,11 @@
 #!/bin/bash
 set -ueo pipefail
 
-BASE64_DECODE_ARGS="-d"
+# BASE64_DECODE_ARGS="-d"
```
Comment (toote) [inline lib/shared.bash:4]: if the code is no longer used, remove it

File: lib/shared.bash [hunk 12]
```diff
@@ -1,11 +1,11 @@
...
 
-case "$(uname -s)" in
-  Darwin) BASE64_DECODE_ARGS="--decode" ;;
-esac
+# case "$(uname -s)" in
+#   Darwin) BASE64_DECODE_ARGS="--decode" ;;
+# esac
```
Comment (toote) [inline lib/shared.bash:8]: same here, if it is no longer used, remove it

File: tests/auth-tests.bats [hunk 13]
```diff
@@ -7,95 +7,46 @@ load '/usr/local/lib/bats/load.bash'
...
+    "kv list -address=https://vault_svr_url -format=yaml foobar : exit 0"
 
   run bash -c "$PWD/hooks/environment && $PWD/hooks/pre-exit"
-
+  
   assert_success
-  assert_output --partial 'Successfully authenticated. You are now logged in'
-
-  unstub vault
+  # assert_output --partial 'Successfully authenticated with RoleID'
```
Comment (toote) [inline tests/auth-tests.bats:28]: why did you make these changes?

* All stubbed commands should be unstubbed to ensure they were called as expected
* you are only testing that the hooks did not fail/explode, but no actual functionality otherwise

File: tests/auth-tests.bats [hunk 14]
```diff
@@ -7,95 +7,46 @@ load '/usr/local/lib/bats/load.bash'
...
 
   unset BUILDKITE_PLUGIN_VAULT_SECRETS_PATH
   unset BUILDKITE_PLUGIN_VAULT_SECRETS_SERVER
   unset BUILDKITE_PLUGIN_VAULT_SECRETS_DUMP_ENV
-  unset BUILDKITE_PLUGIN_VAULT_SECRETS_AUTH_HEADER
   unset BUILDKITE_PIPELINE_SLUG
+  unset BUILDKITE_PLUGIN_VAULT_SECRETS_AUTH_SECRET_ENV
+  unset BUILDKITE_PLUGIN_VAULT_SECRETS_AUTH_ROLE_ID
+  unset BUILDKITE_PLUGIN _VAULT_SECRETS_AUTH_SECRET_ID
+  unset BUILKDITE_PLGUIN_VAULT_SECRETS_AUTH_METHOD
```
Comment (toote) [inline tests/auth-tests.bats:37]: note that it has long been unnecessary to unset all variables exported inside a test, they are all isolated from each other

File: tests/environment-hook.bats [hunk 15]
```diff
@@ -13,14 +13,17 @@ load '/usr/local/lib/bats/load.bash'
...
+  export BUILDKITE_PLUGIN_VAULT_SECRETS_AUTH_ROLE_ID=buildkite
+  export BUILDKITE_PLUGIN_VAULT_SECRETS_AUTH_METHOD=approle
 
   stub vault \
-    'auth -address=https://vault_svr_url - : echo Successfully authenticated. You are now logged in' \
-    "list -address=https://vault_svr_url -format=yaml data/buildkite/testpipe : exit 0 " \
-    "list -address=https://vault_svr_url -format=yaml data/buildkite : echo -e '- env'" \
-    "list -address=https://vault_svr_url -format=yaml data/buildkite/env : echo -e '- MY_SECRET'" \
-    "read -address=https://vault_svr_url -field=value data/buildkite/env/MY_SECRET : echo ${TESTDATA}"
+    "write -field=token -address=https://vault_svr_url auth/approle/login role_id=buildkite secret_id=\"${BUILDKITE_PLUGIN_VAULT_SECRETS_AUTH_SECRET_ENV}\" : echo 'Successfully authenticated with Role ID ${BUILDKITE_PLUGIN_VAULT_SECRETS_AUTH_ROLE_ID}'" \
```
Comment (toote) [inline tests/environment-hook.bats:22]: You can make this stub more generic (and shorter) with the following:

```suggestion
    "write -field=token \* auth/approle/login \* \* : echo 'Successfully authenticated with \${5} \${6}" \
```

Which is basically accepting any parameters in the 3rd, 5th and 6th positions and using the last two in the output so you can later make assertions on the output

File: tests/environment-hook.bats [hunk 16]
```diff
@@ -13,14 +13,17 @@ load '/usr/local/lib/bats/load.bash'
...
+    "kv list -address=https://vault_svr_url -format=yaml data/buildkite/testpipe : echo 'env'" \
+    "kv list -address=https://vault_svr_url -format=yaml data/buildkite : echo 'env'" \
+    "kv get -address=https://vault_svr_url -field=data -format=yaml data/buildkite/testpipe/env : echo ${TESTDATA}" \
+    "kv get -address=https://vault_svr_url -field=data -format=yaml data/buildkite/env : echo ${TESTDATA}"
```
Comment (toote) [inline tests/environment-hook.bats:26]: You are outputting the same string twice, how can you distinguish them in the output? (i.e.: if one of these is not called, the `assert_output` would not fail). I understand that other assertions or the unstub most likely would, but if you are testing some functionality, it may be a good idea to test it all.

File: tests/environment-hook.bats [hunk 17]
```diff
@@ -87,26 +104,30 @@ load '/usr/local/lib/bats/load.bash'
...
   unset BUILDKITE_PIPELINE_SLUG
+  unset BUILDKITE_PLUGIN_VAULT_SECRETS_AUTH_SECRET_ENV
+  unset BUILDKITE_PLUGIN_VAULT_SECRETS_AUTH_ROLE_ID
+  unset BUILDKITE_PLUGIN_VAULT_SECRETS_AUTH_METHOD
 }
 
 #-------
 # Project scope
 @test "Load project env file from vault server" {
+  skip "this test won't pass completely, for now"
```
Comment (toote) [inline tests/environment-hook.bats:120]: can you add more information as to why?

File: tests/environment-hook.bats [hunk 18]
```diff
@@ -123,17 +144,18 @@ load '/usr/local/lib/bats/load.bash'
 }
 
 @test "Load project environment file from vault server" {
+  skip "this test won't pass completely, for now"
```
Comment (toote) [inline tests/environment-hook.bats:147]: can you add more information as to why?

File: tests/environment-hook.bats [hunk 19]
```diff
@@ -150,20 +172,21 @@ load '/usr/local/lib/bats/load.bash'
 }
 
 @test "Load project env and environments files from vault server" {
+  skip "this test won't pass completely, for now"
```
Comment (toote) [inline tests/environment-hook.bats:175]: can you add more information as to why?

**Review by toote — CHANGES_REQUESTED** [review]
Great work! A lot of comments here and there, though.

The most concerning to me are:
* plugins in a single step
* skipping of tests (if you are skipping a test indicate exactly why so that anyone can turn them on afterwards)

File: .buildkite/steps/vault-init-tester.sh [hunk 20]
```diff
@@ -1,16 +1,19 @@
...
-TESTDATA_1=`echo TESTDATA_1=\"foo bar 1\" | base64`
-TESTDATA_2=`echo TESTDATA_2=\"foo bar 2\" | base64`
+vault login token="$AUTH_TOKEN"
+
+[ $? -eq 0 ] && vault secrets enable -path=data/buildkite kv
+
+TESTDATA_1="foobar1"
+TESTDATA_2="foobar2"
 
 [ $? -ne 1 ] && {
```
[the remark this answers is not in the document]
Comment (jeremybumsted) [reply .buildkite/steps/vault-init-tester.sh:16]: @toote I think this is valid since it's not a variable assignment but rather an operation on the vault server. The last command run is `vault login token="$AUTH_TOKEN"` which will authenticate to the vault service that is spun up as part of the integration tests using the `AUTH_TOKEN`, it tests that the script is able to login to `vault` with the supplied token and enable the `data/buildkite` kv2 secrets engine. There might be a more efficient way of doing this though.

File: Dockerfile [hunk 21]
```diff
@@ -0,0 +1,3 @@
+FROM bash:alpine3.16
+
+RUN apk add --no-cache vault
```
[the remark this answers is not in the document]
Comment (jeremybumsted) [reply Dockerfile:3]: Flipped over to using the `vault` image in the Dockerfile and adding bash to the container so we can take advantage of running the tests inside of bash instead of Alpine's default shell

File: .buildkite/steps/test_envvar.sh [hunk 22]
```diff
@@ -1,42 +1,44 @@
...
-  _source="${BASH_SOURCE[0]}"
+  _source="$0"
   [ -z "${_source:-}" ] && _source="${0}"
   basedir="$( cd "$( dirname "${_source}" )" && cd ../.. && pwd )"
 }
 
 setupTestData() {
   export VAULT_TOKEN="${VAULT_DEV_ROOT_TOKEN_ID:-}"
-  ${basedir}/.buildkite/steps/vault-init-tester.sh
+  sh "${basedir}/.buildkite/steps/vault-init-tester.sh"
```
[the remark this answers is not in the document]
Comment (jeremybumsted) [reply .buildkite/steps/test_envvar.sh:13]: Fixed this by switching over to the official `vault` image and adding `bash`

**Comment by jeremybumsted** [comment]
@toote 👋 Ended up going in and just fixing all the tests so everything is working now and passing 👍

File: .buildkite/steps/vault-init-tester.sh (hunk 20 above)
[the remark this answers is not in the document]
Comment (toote) [reply .buildkite/steps/vault-init-tester.sh:16]: What I'm saying is that the value of `$?` (the exit code of the command in line 9) gets overwritten on every single command run. In particular this is the value of the variable for each line:

* line 11: if `vault login` failed or not
* line 13: 
  - if it was `0` in line 11, it is now the result of the `vault secrets enable`
  - otherwise, it will remain unchanged
* line 14: `0` (because the variable assignment of line 13 can not fail)
* live 16: `0` (because the variable assignment of line 14 can not fail)

I double-checked this:
```
$ bash -c 'exit 1'
$ echo $?
1
$ bash -c 'exit 1'
$ [ $? -eq 1 ] && echo true
true
$ echo $?
0
$ bash -c 'exit 1'
$ [ $? -eq 0 ] && echo true
$ echo $?
1
$ bash -c 'exit 1'
$ FOO=10
$ echo $?
0
```

Based on what I believe it is attempted here, and avoiding the use and abuse of `$?` and `test` in the form of `[]`, the script should be something like:

```
if ! vault login ...; then
  echo 'was not able to log in'
  exit 1
fi

if ! vault mount -path data/buildkite kv; then
  echo 'was not able to mount'
  exit 2 # just to be different from the login error
fi

vault kv put data..
```

File: .buildkite/steps/test_envvar.sh [hunk 23]
```diff
@@ -1,42 +1,44 @@
 #!/bin/bash
+
 set -eu
 
 configEnv() {
-  _source="${BASH_SOURCE[0]}"
+  _source="$0"
   [ -z "${_source:-}" ] && _source="${0}"
```
[the remark this answers is not in the document]
Comment (toote) [reply .buildkite/steps/test_envvar.sh:7]: still not sure why line 7 is checking if a variable that you just defined is still defined and if not, defines it the exact same way as you did before

File: .buildkite/steps/test_envvar.sh [hunk 24]
```diff
@@ -1,42 +1,44 @@
...
   basedir="$( cd "$( dirname "${_source}" )" && cd ../.. && pwd )"
 }
 
 setupTestData() {
   export VAULT_TOKEN="${VAULT_DEV_ROOT_TOKEN_ID:-}"
```
[the remark this answers is not in the document]
Comment (toote) [reply .buildkite/steps/test_envvar.sh:12]: still not sure what this export is for

File: .buildkite/steps/test_envvar.sh [hunk 25]
```diff
@@ -1,42 +1,44 @@
...
   export BUILDKITE_PIPELINE_SLUG="${BUILDKITE_PIPELINE_SLUG:-}"
   export BUILDKITE_PLUGIN_VAULT_SECRETS_SERVER="${BUILDKITE_PLUGIN_VAULT_SECRETS_SERVER:-}"
 
-  source ${basedir}/hooks/environment
+  . "${basedir}"/hooks/environment
+
 
-  if [[ ! -z "${TESTDATA_1:-}" ]] && [[ "${TESTDATA_1}" == "foo bar 1" ]] ; then
+  if [[ -n "${TESTDATA_1:-}" ]] && [[ "${TESTDATA_1}" == "foobar1" ]] ; then
```
[the remark this answers is not in the document]
Comment (toote) [reply .buildkite/steps/test_envvar.sh:24]: my previous comment still applies

File: .buildkite/steps/test_envvar.sh [hunk 26]
```diff
@@ -1,42 +1,44 @@
...
     echo "TESTDATA_1 is correct"
   else
     echo "TESTDATA_1 is not set and/or correct"
     exit 1
   fi
 
-  if [[ ! -z "${TESTDATA_2:-}" ]] && [[ "${TESTDATA_2}" == "foo bar 2" ]] ; then
+  if [[ ! -z "${TESTDATA_2:-}" ]] && [[ "${TESTDATA_2}" == "foobar2" ]] ; then
```
[the remark this answers is not in the document]
Comment (toote) [reply .buildkite/steps/test_envvar.sh:31]: my previous comment still applies here

File: .buildkite/steps/vault-init-tester.sh [hunk 27]
```diff
@@ -1,16 +1,19 @@
 #!/bin/bash
 
-# This token is defined in docker-compoise.yml and is for testing ONLY!!!!!!
+# This token is defined in docker-compose.yml and is for testing ONLY!!!!!!
 AUTH_TOKEN="${BUILDKITE_PLUGIN_VAULT_SECRETS_AUTH_TOKEN:-}"
 PROJECT=${BUILDKITE_PIPELINE_SLUG:-foobar_project}
-vault auth - <<< "$AUTH_TOKEN"
+export VAULT_ADDR="http://vault-svc:8200"
```
[the remark this answers is not in the document]
Comment (toote) [reply .buildkite/steps/vault-init-tester.sh:6]: my previous comment still applies here

File: tests/environment-hook.bats [hunk 28]
```diff
@@ -29,25 +31,22 @@ load '/usr/local/lib/bats/load.bash'
...
+  export BUILDKITE_PLUGIN_VAULT_SECRETS_AUTH_ROLE_ID=buildkite
+  export BUILDKITE_PLUGIN_VAULT_SECRETS_AUTH_METHOD=approle
 
   stub vault \
-    'auth -address=https://vault_svr_url - : echo Successfully authenticated. You are now logged in' \
-    "list -address=https://vault_svr_url -format=yaml data/buildkite/testpipe : exit 0 "\
-    "list -address=https://vault_svr_url -format=yaml data/buildkite : echo -e '- environment'" \
-    "list -address=https://vault_svr_url -format=yaml data/buildkite/environment : echo -e '- MY_SECRET'" \
-    "read -address=https://vault_svr_url -field=value data/buildkite/environment/MY_SECRET : echo ${TESTDATA}"
+    "write -field=token -address=https://vault_svr_url auth/approle/login role_id=buildkite secret_id=\"${BUILDKITE_PLUGIN_VAULT_SECRETS_AUTH_SECRET_ENV}\" : echo 'Successfully authenticated with Role ID ${BUILDKITE_PLUGIN_VAULT_SECRETS_AUTH_ROLE_ID}'" \
```
Comment (toote) [inline tests/environment-hook.bats:46]: you can reuse the same generic stub as before to make it more general and less verbose:

```suggestion
    "write -field=token \* auth/approle/login \* \* : echo 'Successfully authenticated with \${5} \${6}" \
```

File: tests/environment-hook.bats [hunk 29]
```diff
@@ -56,28 +55,26 @@ load '/usr/local/lib/bats/load.bash'
...
 
   stub vault \
-    'auth -address=https://vault_svr_url - : echo Successfully authenticated. You are now logged in' \
-    "list -address=https://vault_svr_url -format=yaml data/buildkite/testpipe : exit 0 "\
-    "list -address=https://vault_svr_url -format=yaml data/buildkite : echo -e '- env\n- environment'" \
-    "list -address=https://vault_svr_url -format=yaml data/buildkite/env : echo -e '- MY_SECRET'" \
-    "read -address=https://vault_svr_url -field=value data/buildkite/env/MY_SECRET : echo ${TESTDATA_ENV1}" \
-    "list -address=https://vault_svr_url -format=yaml data/buildkite/environment : echo -e '- ANOTHER_SECRET'" \
-    "read -address=https://vault_svr_url -field=value data/buildkite/environment/ANOTHER_SECRET : echo ${TESTDATA_ENV2}"
+    "write -field=token -address=https://vault_svr_url auth/approle/login role_id=buildkite secret_id=\"${BUILDKITE_PLUGIN_VAULT_SECRETS_AUTH_SECRET_ENV}\" : echo 'Successfully authenticated with Role ID ${BUILDKITE_PLUGIN_VAULT_SECRETS_AUTH_ROLE_ID}'" \
```
Comment (toote) [inline tests/environment-hook.bats:73]: same here:
```suggestion
    "write -field=token \* auth/approle/login \* \* : echo 'Successfully authenticated with \${5} \${6}" \
```

**Review by toote — CHANGES_REQUESTED** [review]
Still several comments of my previous review remain unaddressed and I found a few more things to check. Finally, the tests are still not passing on the latest plugin tester version (v4.0.0), luckily it is a simple fix in the ssh tests (missing a `*` in the shh-keygen stub of the first test)

File: .buildkite/steps/vault-init-tester.sh (hunk 27 above)
[the remark this answers is not in the document]
Comment (jeremybumsted) [reply .buildkite/steps/vault-init-tester.sh:6]: This one is a bit weird... I've tried a couple of different methods to make sure things work appropriately, but basically if no value is set, `VAULT_ADDR` will default to `http://localhost:8200` which will cause the integration tests to fail. I've managed to work around this though, so this is fixed.

File: tests/auth-tests.bats (hunk 13 above)
Comment (jeremybumsted) [reply tests/auth-tests.bats]: These and the other tests are fixed

File: tests/auth-tests.bats (hunk 14 above)
Comment (jeremybumsted) [reply tests/auth-tests.bats]: TIL! Updated to remove the unsetting in all the tests 🙌

File: .buildkite/steps/test_envvar.sh (hunk 22 above)
[the remark this answers is not in the document]
Comment (jeremybumsted) [reply .buildkite/steps/test_envvar.sh:13]: Ah, this was a mistake 😓 I switched this up trying to get the script to run in the integration-test container, I've reverted back to running in the same shell now.

File: .buildkite/steps/vault-init-tester.sh (hunk 20 above)
[the remark this answers is not in the document]
Comment (jeremybumsted) [reply .buildkite/steps/vault-init-tester.sh:16]: Ahhh, this is where my knowledge of bash is pretty weak so I've learned some new things here. I've tested your suggestions and things are working as expected, thanks for that explanation!

File: .buildkite/steps/test_envvar.sh (hunk 23 above)
[the remark this answers is not in the document]
Comment (jeremybumsted) [reply .buildkite/steps/test_envvar.sh:7]: Yeah, doesn't make much sense... I've updated it as such:

```shell
configEnv() {
  basedir="$( cd "$( dirname "$0" )" && cd ../.. && pwd )"
}
```

File: .buildkite/steps/test_envvar.sh (hunk 24 above)
[the remark this answers is not in the document]
Comment (jeremybumsted) [reply .buildkite/steps/test_envvar.sh:12]: Yeah, not needed - removed.

File: tests/environment-hook.bats (hunk 15 above)
Comment (jeremybumsted) [reply tests/environment-hook.bats]: I've actually completely removed the auth section of these tests because it's not needed to test the functionality of evaluating `env` received from Vault. This is good to know though!

File: .buildkite/pipeline.yml [hunk 30]
```diff
@@ -0,0 +1,34 @@
...
+    plugins:
+      - plugin-linter#v2.0.0:
+          id: vault-secrets
+  - label: ":test_tube: Tests"
+    plugins:
+      - docker-compose#v4.5.0:
+          run: tests
+          
+  - label: ":vault: :test_tube: Integration Tests"
+    command: bash .buildkite/steps/test_envvar.sh
```
[the remark this answers is not in the document]
Comment (jeremybumsted) [reply .buildkite/pipeline.yml:21]: This was an artifact of the old images I was using, no longer needed in the current configuration, so removed

File: tests/environment-hook.bats (hunk 28 above)
Comment (jeremybumsted) [reply tests/environment-hook.bats]: Both this and the other tests which had an auth method [configured ](https://github.com/buildkite-plugins/vault-secrets-buildkite-plugin/pull/13#discussion_r1111105758) didn't actually need them to be defined to test the functionality, so I've removed this stub

**Comment by jeremybumsted** [comment]
@toote Thanks again for the review! I've addressed your comments, as well as fixed some unintended behaviour in the auth function that wasn't using the secret ID properly (and causing some tests to fail inadvertently).

**Review by toote — APPROVED** [review]
:clap: awesome work!
